provider/oracle/environ.go

```diff
@@ -0,0 +1,749 @@
+	}
+
+	env := &oracleEnviron{
+		p:      p,
+		spec:   args.Cloud,
+		cfg:    args.Config,
+		mutex:  &sync.Mutex{},
+		client: client,
+	}
+	// create a new firewall from the env and the internal api client
+	env.firewall = NewFirewall(env, client)
+
+	namespace, err := instance.NewNamespace(env.cfg.UUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	env.namespace = namespace
+
+	return env, nil
+}
+
+// PrepareForBootstrap prepares an environment for bootstrapping.
+//
+// This will be called very early in the bootstrap procedure, to
+// give an Environ a chance to perform interactive operations that
+// are required for bootstrapping.
+func (o *oracleEnviron) PrepareForBootstrap(ctx environs.BootstrapContext) error {
+	if ctx.ShouldVerifyCredentials() {
+		logger.Infof("Loging into the oracle cloud infrastructure")
+		if err := o.client.Authenticate(); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+// Bootstrap creates a new environment, and an instance inside the
+// oracle cloud infrastracture to host the controller for that
+// environment. The instnace will have have the series and architecture
+// of the Environ's choice, constrained to those of the available tools.
+// Bootstrap will return the instance's
+// architecture, series, and a function that must be called to finalize
+// the bootstrap process by transferring the tools and installing the
+// initial Juju controller.
+func (o *oracleEnviron) Bootstrap(
+	ctx environs.BootstrapContext,
+	args environs.BootstrapParams,
+) (*environs.BootstrapResult, error) {
+
+	return common.Bootstrap(ctx, o, args)
+}
+
+// Create creates the environment for a new hosted model.
+//
+// This will be called before any workers begin operating on the
+// Environ, to give an Environ a chance to perform operations that
+// are required for further use.
+//
+// Create is not called for the initial controller model; it is
+// the Bootstrap method's job to create the controller model.
+func (o *oracleEnviron) Create(params environs.CreateParams) error {
+	if err := o.client.Authenticate(); err != nil {
+		return errors.Trace(err)
+	}
+
+	return nil
+}
+
+// AdoptResources is called when the model is moved from one
+// controller to another using model migration. Some providers tag
+// instances, disks, and cloud storage with the controller UUID to
+// aid in clean destruction. This method will be called on the
+// environ for the target controller so it can update the
+// controller tags for all of those things. For providers that do
+// not track the controller UUID, a simple method returning nil
+// will suffice. The version number of the source controller is
+// provided for backwards compatibility - if the technique used to
+// tag items changes, the version number can be used to decide how
+// to remove the old tags correctly.
+func (e *oracleEnviron) AdoptResources(controllerUUID string, fromVersion version.Number) error {
+	return nil
```

Comment:
Please leave a TODO to implement this in a follow-up. This is required for model migration to work properly, as mentioned in the comment.